lib/config/flat-config-schema.js

```diff
@@ -0,0 +1,362 @@
+        throw new TypeError("Expected a string, number, or array.");
+    }
+}
+
+/**
+ * Validates that a value is valid rule severity.
+ * @param {any} value The value to check.
+ * @returns {void}
+ * @throws {TypeError} If the value isn't a valid rule severity.
+ */
+function assertIsRuleSeverity(value) {
+    const severity = typeof value === "string"
+        ? ruleSeverities.get(value.toLowerCase())
+        : ruleSeverities.get(value);
+
+    if (typeof severity === "undefined") {
+        throw new TypeError("Expected severity of \"off\", 0, \"warn\", 1, \"error\", or 2.");
+    }
+}
+
+
+/**
+ * Validates that a given string is the form pluginName/objectName.
+ * @param {string} value The string to check.
+ * @returns {void}
+ * @throws {TypeError} If the string isn't in the correct format.
+ */
+function assertIsPluginMemberName(value) {
+    if (!/[@a-z0-9-_$]+\/(?:[a-z0-9-_$]+)+/iu.test(value)) {
+        throw new TypeError(`Expected string in the form "pluginName/objectName" but found "${value}".`);
+    }
+}
+
+/**
+ * Validates that a value is an object.
+ * @param {any} value The value to check.
+ * @returns {void}
+ * @throws {TypeError} If the value isn't an object.
+ */
+function assertIsObject(value) {
+    if (!value || typeof value !== "object") {
+        throw new TypeError("Expected an object.");
+    }
+}
+
+/**
+ * Validates that a value is an object or a string.
+ * @param {any} value The value to check.
+ * @returns {void}
+ * @throws {TypeError} If the value isn't an object or a string.
+ */
+function assertIsObjectOrString(value) {
+    if ((!value || typeof value !== "object") && typeof value !== "string") {
+        throw new TypeError("Expected an object or string.");
+    }
+}
+
+//-----------------------------------------------------------------------------
+// Low-Level Schemas
+//-----------------------------------------------------------------------------
+
+/** @type {ObjectPropertySchema} */
+const objectAssignSchema = {
+    merge: "assign",
+    validate: "object"
+};
+
+/** @type {ObjectPropertySchema} */
+const numberSchema = {
+    merge: "replace",
+    validate: "number"
+};
+
+/** @type {ObjectPropertySchema} */
+const booleanSchema = {
+    merge: "replace",
+    validate: "boolean"
+};
+
+//-----------------------------------------------------------------------------
+// High-Level Schemas
+//-----------------------------------------------------------------------------
+
+/** @type {ObjectPropertySchema} */
+const globalsSchema = {
+    merge: "assign",
+    validate(value) {
+
+        assertIsObject(value);
+
+        for (const key of Object.keys(value)) {
+
+            // avoid hairy edge case
+            if (key === "__proto__") {
+                continue;
+            }
+
+            if (key !== key.trim()) {
+                throw new TypeError(`Global "${key}" has leading or trailing whitespace.`);
```

Comment:
There's not currently a test exercising this check.

Reply:
Added